File: src/main/java/cat/xlagunas/andrtc/token/TokenRepositoryImpl.java
package cat.xlagunas.andrtc.token;

import org.springframework.dao.DuplicateKeyException;
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.jdbc.support.KeyHolder;

import java.util.List;

public class TokenRepositoryImpl implements TokenRepository {
    private final static String INSERT_TOKEN = "INSERT INTO TOKEN (OWNER, TOKEN, PLATFORM) VALUES (:owner, :token, :platform)";
    private final static String FIND_USER_TOKENS = "SELECT * FROM TOKEN WHERE TOKEN.OWNER = :owner";
    private final static String FIND_LIST_USER_TOKENS = "SELECT * FROM TOKEN WHERE TOKEN.OWNER IN (:ownerList)";
    private final static String DELETE_TOKEN = "DELETE FROM TOKEN WHERE TOKEN.OWNER = :owner AND TOKEN.TOKEN = :token";

    NamedParameterJdbcTemplate jdbcTemplate;

    public TokenRepositoryImpl(NamedParameterJdbcTemplate template) {
        this.jdbcTemplate = template;
    }

    @Override
    public long addToken(long userId, String token, String platform) throws ExistingTokenException {
        try {
            KeyHolder holder = new GeneratedKeyHolder();
            long id = jdbcTemplate.update(INSERT_TOKEN,
                    new MapSqlParameterSource("owner", userId)
                            .addValue("token", token)
                            .addValue("platform", platform),
                    holder);
            return holder.getKey().longValue();
        } catch (DuplicateKeyException ex) {
            throw new ExistingTokenException(ex);
        }
    }

    @Override
    public List<Token> getUserTokens(long userId) {
        return jdbcTemplate.query(FIND_USER_TOKENS, new MapSqlParameterSource("owner", userId), TokenRowMapper.insertMapper());
    }

    public List<Token> getUserListToken(List<Long> userId) {
        return jdbcTemplate.query(FIND_LIST_USER_TOKENS, new MapSqlParameterSource("ownerList", userId), TokenRowMapper.insertMapper());
    }

    @Override
    public boolean removeToken(long userId, String token) {
        int affectedRows = jdbcTemplate.update(DELETE_TOKEN,
                new MapSqlParameterSource("owner", userId)
                        .addValue("token", token));
        return affectedRows > 0;
    }
}

File: src/main/java/cat/xlagunas/andrtc/token/TokenRepositoryImpl.kt
package cat.xlagunas.andrtc.token

import org.springframework.dao.DuplicateKeyException
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.jdbc.support.GeneratedKeyHolder

class TokenRepositoryImpl(internal var jdbcTemplate: NamedParameterJdbcTemplate) : TokenRepository {

    @Throws(ExistingTokenException::class)
    override fun addToken(userId: Long, token: String, platform: String): Long {
        try {
            val queryParams = MapSqlParameterSource()
                    .addValue("owner", userId)
                    .addValue("token", token)
                    .addValue("platform", platform)

            return jdbcTemplate.update(INSERT_TOKEN, queryParams, GeneratedKeyHolder()).toLong()
        } catch (ex: DuplicateKeyException) {
            throw ExistingTokenException(ex)
        }
    }

    override fun getUserTokens(userId: Long): List<Token> {
        val queryParams = MapSqlParameterSource("owner", userId)
        return jdbcTemplate.query(FIND_USER_TOKENS, queryParams, TokenRowMapper.insertMapper())
    }

    override fun getUserListToken(userId: List<Long>): List<Token> {
        val queryParams = MapSqlParameterSource("ownerList", userId)
        return jdbcTemplate.query(FIND_LIST_USER_TOKENS, queryParams, TokenRowMapper.insertMapper())
    }

    override fun removeToken(userId: Long, token: String): Boolean {
        val queryParams = MapSqlParameterSource()
                .addValue("owner", userId)
                .addValue("token", token)
        return jdbcTemplate.update(DELETE_TOKEN, queryParams) > 0
    }

    companion object {
        private val INSERT_TOKEN = "INSERT INTO TOKEN (OWNER, TOKEN, PLATFORM) VALUES (:owner, :token, :platform)"
        private val FIND_USER_TOKENS = "SELECT * FROM TOKEN WHERE TOKEN.OWNER = :owner"
        private val FIND_LIST_USER_TOKENS = "SELECT * FROM TOKEN WHERE TOKEN.OWNER IN (:ownerList)"
        private val DELETE_TOKEN = "DELETE FROM TOKEN WHERE TOKEN.OWNER = :owner AND TOKEN.TOKEN = :token"
    }
}
